### ISSUE TITLE
Embed tarantool/checks in core tarantool

### ISSUE THREAD

**Issue by DifferentialOrange** [tarantool/tarantool#7726]
The series of tickets about tarantool/metrics embedding is just a beginning of the discussion started with possible Product Team plans related to Flight Recorder development. Expected features may be reevaluated, estimation is preliminary.

[tarantool/checks](https://github.com/tarantool/checks) is a Lua module that declares a `checks()` function and a `checkers` table, which allow to check the parameters passed to a Lua function in a fast and unobtrusive way.

Embedding tarantool/checks is a prerequirement to embedding tarantool/metrics (see #7725).

Disregarding #7725, since `tarantool/checks` is a language extension, it seems like a good idea to provide it together with our language implementation.

We also need to update module features since it doesn't support 2.10 types: datetime and interval.

Optimization of module implementation is also may be considered since using it in function always has affected performance. As part of the solution, handles for checks disable could be implemented (similar to https://github.com/tarantool/crud/blob/f1b6a12188f3cd4278fb879afc676b4cd5b5f3ec/crud/common/dev_checks.lua).

**Definition of Done**

Minimal: any application that uses the module can use `require('checks')` and another features without any additional rocks installation.

Additional: performance optimization.

**Comment by DifferentialOrange** [comment]
Current task estimation is based on the following solution:
- supporting 2.10 types,
- embedding the module,
- porting tests.

**Comment by Totktonada** [comment]
There is also very similar issue #3053 (but here we have more precise action plan).

**Comment by DifferentialOrange** [comment]
The other question is the life after embedding `checks`. It seems that when the module is embedded, it becomes nearly impossible to load the non-builtin version of the `checks` (at least with common approaches). For example, it means that `reusable_testing` pipeline became useless since it will test built-in checks. The main issue is that after embedding updating `checks` for users will be tied to updating core Tarantool binary.

In case of embedding `metrics`, we've [decided to create a separate rock `metrics-extentions`](https://github.com/tarantool/tarantool/issues/7727) to solve this issue. All updates of `metrics` code will be put through this rock, so people with non-bleeding Tarantool binaries will be able to setup new features and bugfixes. What we should do with `checks`?

Compared to `metrics`, `checks` is a small module (a single file with ~350 lines of Lua code) and it's rarely updated (the code had a single commit in 2020, another single commit in 2021 and one more single commit in 2022). It is rather widely used in Tarantool modules and applications, and for now it hasn't any known bugs. There are also no feature request tickets, and it is expected that all future `checks` feature requests will be related to adding new types in core Tarantool, so it is unlikely that there will be any demand to use newest `checks` with old Tarantool.

(After expecting `checks` code, I've found a single bug related to LuaJit version and a group of missing types (`error`, `datetime`, `interval`), which is planned to be covered by https://github.com/tarantool/checks/pull/42 PR which is the part of this ticket.)

One more approach to this issue is to embed `checks` depending on some flag. It would allow to build Tarantool without in-built `checks`, but things become much complicated here. Which should be used by default? If default is "build with checks", than this flag would seem to be rather useless since the absolute majority of our customers uses pre-build Enterprise SDK or Community packages. If default if "build without checks", then we'll never solve the original issue to collect metrics on board since, again, the absolute majority of our customers will use Tarantool with checks (and metrics) disabled on build.

So, **the proposed solution** is "it seems to be the best idea to unconditionally embed `checks` to each new Tarantool binary and introduce any updates with commits to the original repo and bumping the commit in the core repo".

**Comment by Totktonada** [comment]
I think that it is better to copy the module code into tarantool sources. This way we'll be able to provide a fast Lua/C implementation in tarantool and easy-to-distribute pure Lua implementation as a separate module.

It's completely ok to leave the separate module existing: IOW, there is no reason to do anything with it. However after #7774 we can enhance the external module with ability to replace the built-in one. It'll allow to deliver a bugfix withup updating tarantool by shipping the new version of the external module.

I suggest to add a `_VERSION` field into the built-in and the external modules as soon as possible (https://github.com/tarantool/roadmap-internal/issues/204). It'll allow to choose the best version later, when the built-in and the external ones are available.

**Comment by DifferentialOrange** [comment]
> I suggest to add a `_VERSION` field into the built-in and the external modules as soon as possible ([tarantool/roadmap-internal#204](https://github.com/tarantool/roadmap-internal/issues/204)). It'll allow to choose the best version later, when the built-in and the external ones are available.

To be honest, I'm not completely understand this one. What do we expect in-built `checks` `_VERSION` to be?

**Comment by Totktonada** [comment]
> To be honest, I'm not completely understand this one. What do we expect in-built `checks` `_VERSION` to be?

A string that represents a version of the module in the semver or a semver alike format. I guess that it'll be `X.Y.Z` for all or at least most modules: we don't need alpha/beta/rc and other semver features here.

**Comment by DifferentialOrange** [comment]
> > To be honest, I'm not completely understand this one. What do we expect in-built `checks` `_VERSION` to be?
> 
> A string that represents a version of the module in the semver or a semver alike format. I guess that it'll be `X.Y.Z` for all or at least most modules: we don't need alpha/beta/rc and other semver features here.

So if we implement something like https://github.com/tarantool/graphql/pull/53/files in tarantool/checks, embedded `checks.lua` will have hardcoded `VERSION` equal to the tag it based on?

**Comment by DifferentialOrange** [comment]
As the result of discussions, for now we decided to embed `checks` as a submodule since it is expected to be simpler to support and we don't plan to drop the support of a separate rock anytime soon. Providing `_VERSION` is a separate task, and since there are no usage in core for it now, we decided not to treat it as a part of this ticket.

Regarding overloading, there is actually a way to do it and soon we'll get a more native way to overload it (which is also not treated as a part of this ticket).
